File: app/src/main/java/com/friendlypos/preventas/adapters/PrevResumenAdapter.java
package com.friendlypos.preventas.adapters;


import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.friendlypos.R;
import com.friendlypos.distribucion.modelo.Pivot;
import com.friendlypos.preventas.activity.PreventaActivity;
import com.friendlypos.preventas.fragment.PrevResumenFragment;
import com.friendlypos.preventas.modelo.Bonuses;
import com.friendlypos.principal.modelo.Productos;

import java.util.ArrayList;
import java.util.List;

import io.realm.Realm;

/**
 * Created by DelvoM on 21/09/2017.
 */

public class PrevResumenAdapter extends RecyclerView.Adapter<PrevResumenAdapter.CharacterViewHolder> {
    private List<Pivot> productosList;
    private PreventaActivity activity;
    private static ArrayList<Pivot> aListdata = new ArrayList<Pivot>();
    private int selected_position1 = -1;
    Double amount_inventario = 0.0;
    int idInvetarioSelec;
    int nextId;
    private PrevResumenFragment fragment;
    private static double productosDelBonus = 0;

    public PrevResumenAdapter(PreventaActivity activity, PrevResumenFragment fragment, List<Pivot> productosList) {
        this.productosList = productosList;
        this.activity = activity;
        this.fragment = fragment;
    }

    public PrevResumenAdapter() {

    }

    public void updateData(List<Pivot> productosList) {
        activity.cleanTotalize();
        this.productosList = productosList;
        notifyDataSetChanged();

    }

    @Override
    public CharacterViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.lista_distribucion_resumen, parent, false);
        return new PrevResumenAdapter.CharacterViewHolder(view);

    }

    private String getProductDescriptionByPivotId(String id) {
        Realm realm = Realm.getDefaultInstance();
        String description = realm.where(Productos.class).equalTo("id", id).findFirst().getDescription();
        realm.close();
        return description;
    }

    @Override
    public void onBindViewHolder(PrevResumenAdapter.CharacterViewHolder holder, final int position) {

        final Pivot pivot = productosList.get(position);
        String pivotTotal;
        double amountBonif;
        holder.txt_resumen_factura_nombre.setText(getProductDescriptionByPivotId(pivot.getProduct_id()));
        holder.txt_resumen_factura_precio.setText("P: " + Double.valueOf(pivot.getPrice()));
        holder.txt_resumen_factura_descuento.setText("Descuento de: " + Double.valueOf(pivot.getDiscount()));
        holder.txt_resumen_factura_cantidad.setText("C: " + Double.parseDouble(pivot.getAmount()));

            Realm realm0 = Realm.getDefaultInstance();
            String bonus = realm0.where(Productos.class).equalTo("id", pivot.getProduct_id()).findFirst().getBonus();
            realm0.close();

        if (bonus.equals("1")){

            final Realm realmBonus = Realm.getDefaultInstance();

            realmBonus.executeTransaction(new Realm.Transaction() {

                @Override
                public void execute(Realm realmBonus) {

                    Bonuses productoConBonus = realmBonus.where(Bonuses.class).equalTo("product_id", Integer.valueOf(pivot.getProduct_id())).findFirst();
                    productosDelBonus = Double.parseDouble(productoConBonus.getProduct_bonus());

                    Log.d("BONIFTOTAL", productoConBonus.getProduct_id() +  " " + productosDelBonus);

                }
            });
            amountBonif = Double.parseDouble(pivot.getAmount()) - productosDelBonus;
            pivotTotal = String.format("%,.2f", (Double.valueOf(pivot.getPrice()))* amountBonif);
            Log.d("BONIFTOTALPIVOT", pivotTotal +  " ");
        }
        else{
            pivotTotal = String.format("%,.2f", (Double.valueOf(pivot.getPrice()) * Double.parseDouble(pivot.getAmount())));
            Log.d("TOTALPIVOT", pivotTotal +  " ");
        }

        holder.txt_resumen_factura_total.setText("T: " + pivotTotal);

    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public int getItemCount() {
        return productosList.size();
    }

    @Override
    public int getItemViewType(int position) {
        return 0;
    }


    public class CharacterViewHolder extends RecyclerView.ViewHolder {

        private TextView txt_resumen_factura_nombre, txt_resumen_factura_descuento, txt_resumen_factura_precio, txt_resumen_factura_cantidad, txt_resumen_factura_total;
        protected CardView cardView;
        ImageButton btnEliminarResumen;

        public CharacterViewHolder(View view) {
            super(view);
            cardView = (CardView) view.findViewById(R.id.cardViewResumen);
            txt_resumen_factura_nombre = (TextView) view.findViewById(R.id.txt_resumen_factura_nombre);
            txt_resumen_factura_descuento = (TextView) view.findViewById(R.id.txt_resumen_factura_descuento);
            txt_resumen_factura_precio = (TextView) view.findViewById(R.id.txt_resumen_factura_precio);
            txt_resumen_factura_cantidad = (TextView) view.findViewById(R.id.txt_resumen_factura_cantidad);
            txt_resumen_factura_total = (TextView) view.findViewById(R.id.txt_resumen_factura_total);
            btnEliminarResumen = (ImageButton) view.findViewById(R.id.btnEliminarResumen);

            btnEliminarResumen.setOnClickListener(new View.OnClickListener() {

                @Override
                public void onClick(View view) {

                    activity.cleanTotalize();
                    int pos = getAdapterPosition();

                    notifyItemChanged(selected_position1);
                    selected_position1 = getAdapterPosition();
                    notifyItemChanged(selected_position1);

                    final Pivot clickedDataItem = productosList.get(pos);

                    final int resumenProductoId = clickedDataItem.getId();
                    final double cantidadProducto = Double.parseDouble(clickedDataItem.getAmount());

                    Toast.makeText(view.getContext(), "Se borró el producto" , Toast.LENGTH_SHORT).show();

                    // TRANSACCIÓN BD PARA BORRAR EL CAMPO

                    activity.initProducto(pos);

                    notifyDataSetChanged();
                    fragment.updateData();

                }
            });
        }

    }
    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
    }


}

File: app/src/main/java/com/friendlypos/preventas/adapters/PrevResumenAdapter.kt
package com.friendlysystemgroup.friendlypos.preventas.adapters

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import android.widget.Toast
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.friendlysystemgroup.friendlypos.R
import com.friendlysystemgroup.friendlypos.distribucion.modelo.Pivot
import com.friendlysystemgroup.friendlypos.preventas.activity.PreventaActivity
import com.friendlysystemgroup.friendlypos.preventas.fragment.PrevResumenFragment
import com.friendlysystemgroup.friendlypos.preventas.modelo.Bonuses
import com.friendlysystemgroup.friendlypos.principal.modelo.Productos
import io.realm.Realm


/**
 * Created by DelvoM on 21/09/2017.
 */
class PrevResumenAdapter : RecyclerView.Adapter<PrevResumenAdapter.CharacterViewHolder> {
    private var productosList: List<Pivot>? = null
    private var activity: PreventaActivity? = null
    private var selected_position1 = -1
    var amount_inventario: Double = 0.0
    var idInvetarioSelec: Int = 0
    var nextId: Int = 0
    private var fragment: PrevResumenFragment? = null

    constructor(
        activity: PreventaActivity?,
        fragment: PrevResumenFragment?,
        productosList: List<Pivot>?
    ) {
        this.productosList = productosList
        this.activity = activity
        this.fragment = fragment
    }

    constructor()

    fun updateData(productosList: List<Pivot>?) {
        activity!!.cleanTotalize()
        this.productosList = productosList
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CharacterViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.lista_distribucion_resumen, parent, false)
        return CharacterViewHolder(view)
    }

    private fun getProductDescriptionByPivotId(id: String?): String {
        val realm = Realm.getDefaultInstance()
        val description = realm.where(Productos::class.java)
            .equalTo("id", id).findFirst()!!
            .description
        realm.close()
        return description
    }

    override fun onBindViewHolder(holder: CharacterViewHolder, position: Int) {
        val pivot = productosList!![position]
        val pivotTotal: String
        val amountBonif: Double
        holder.txt_resumen_factura_nombre.text = getProductDescriptionByPivotId(pivot.product_id)
        holder.txt_resumen_factura_precio.text = "P: " + pivot.price!!.toDouble()
        holder.txt_resumen_factura_descuento.text = "Descuento de: " + pivot.discount!!.toDouble()
        holder.txt_resumen_factura_cantidad.text = "C: " + pivot.amount!!.toDouble()

        val realm0 = Realm.getDefaultInstance()
        val bonus = realm0.where(Productos::class.java)
            .equalTo("id", pivot.product_id).findFirst()!!
            .bonus
        realm0.close()

        if (bonus == "1") {
            val realmBonus = Realm.getDefaultInstance()

            realmBonus.executeTransaction { realmBonus ->
                val productoConBonus = realmBonus.where(Bonuses::class.java)
                    .equalTo("product_id", pivot.product_id!!.toInt()).findFirst()
                productosDelBonus =
                    productoConBonus!!.product_bonus.toDouble()
                Log.d(
                    "BONIFTOTAL",
                    productoConBonus.product_id.toString() + " " + productosDelBonus
                )
            }
            amountBonif = pivot.amount!!.toDouble() - productosDelBonus
            pivotTotal = String.format("%,.2f", (pivot.price!!.toDouble()) * amountBonif)
            Log.d("BONIFTOTALPIVOT", "$pivotTotal ")
        } else {
            pivotTotal =
                String.format("%,.2f", (pivot.price!!.toDouble() * pivot.amount!!.toDouble()))
            Log.d("TOTALPIVOT", "$pivotTotal ")
        }

        holder.txt_resumen_factura_total.text = "T: $pivotTotal"
    }

    override fun getItemId(position: Int): Long {
        return 0
    }

    override fun getItemCount(): Int {
        return productosList!!.size
    }

    override fun getItemViewType(position: Int): Int {
        return 0
    }


    inner class CharacterViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val txt_resumen_factura_nombre: TextView =
            view.findViewById<View>(R.id.txt_resumen_factura_nombre) as TextView
        val txt_resumen_factura_descuento: TextView =
            view.findViewById<View>(R.id.txt_resumen_factura_descuento) as TextView
        val txt_resumen_factura_precio: TextView =
            view.findViewById<View>(R.id.txt_resumen_factura_precio) as TextView
        val txt_resumen_factura_cantidad: TextView =
            view.findViewById<View>(R.id.txt_resumen_factura_cantidad) as TextView
        val txt_resumen_factura_total: TextView =
            view.findViewById<View>(R.id.txt_resumen_factura_total) as TextView
        protected var cardView: CardView =
            view.findViewById<View>(R.id.cardViewResumen) as CardView
        var btnEliminarResumen: ImageButton =
            view.findViewById<View>(R.id.btnEliminarResumen) as ImageButton

        init {
            btnEliminarResumen.setOnClickListener { view ->
                activity!!.cleanTotalize()
                val pos = adapterPosition

                notifyItemChanged(selected_position1)
                selected_position1 = adapterPosition
                notifyItemChanged(selected_position1)

                val clickedDataItem = productosList!![pos]

                val resumenProductoId = clickedDataItem.id
                val cantidadProducto = clickedDataItem.amount!!.toDouble()

                Toast.makeText(view.context, "Se borró el producto", Toast.LENGTH_SHORT).show()

                // TRANSACCIÓN BD PARA BORRAR EL CAMPO
                activity!!.initProducto(pos)

                notifyDataSetChanged()
                fragment!!.updateData()
            }
        }
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
    }


    companion object {
        private val aListdata = ArrayList<Pivot>()
        private var productosDelBonus = 0.0
    }
}

